### PR TITLE
Server Testkit

### PR DESCRIPTION
**Pull request by andrefmrocha**
Added a Testkit for the implementations of the server. The idea is to create a Mock Client in order to test several BSP features.
As of right now I am testing how it works against my [bsp implementation in bazel](https://github.com/andrefmrocha/bazel-bsp).
I also plan to add tests against the `HappyMockServer`.

### CODE REVIEW

File: testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+import scala.collection.convert.ImplicitConversions.`collection asJava`
+import scala.collection.mutable
+import scala.compat.java8.FutureConverters._
+import scala.compat.java8.OptionConverters._
+import scala.concurrent.Await
+import scala.concurrent.duration.{Duration, DurationInt}
+import scala.jdk.CollectionConverters._
+import scala.util.{Failure, Try}
+
+case class TestClient(
```
Comment (jastice) [inline testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala:20]: case class not a good fit here, it usually implies that it's an immutable data holder, whereas this class kind of holds state

Comment (andrefmrocha) [reply testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala]: Thanks for noticing this. The initial plan for the class was different but I didn't remember to change this

File: testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+  }
+
+  private def testShutdown(): Unit = {
+    await(server.buildShutdown())
+    val failedRequest =
+      Await.ready(server.workspaceBuildTargets().toScala, timeoutDuration).value.get
+    assert(failedRequest.isFailure, "Server is still accepting requests after shutdown")
+  }
+
+  private def wrapTest(body: () => Unit): Unit = {
```
[the remark this answers is not in the document]
Comment (andrefmrocha) [reply testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala]: I had to change this to take a parameter. I believe it unfortunately invalidates this doesn't it?

File: testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+    workspacePath: String,
+    compilerOutputDir: String,
+    session: MockSession,
+    timeoutDuration: Duration = 30.seconds
+) {
+  private val server = session.connection.server
+  private val client = session.client
+  private val alreadySentDiagnosticsTimeout = 2.seconds
+
+  private val unitTests: Map[TestClient.ClientUnitTest.Value, () => Unit] = Map(
```
[the remark this answers is not in the document]
Comment (andrefmrocha) [reply testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala]: Every initialization has to have a shutdown afterwards, doesn't this mean that only a test case for both at the same time could be exposed. Am I wrong?

[the remark this answers is not in the document]
Comment (jastice) [reply testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala]: Yes, makes sense. That would be the "minim viable lifecycle" test :)

File: testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+  }
+
+  def testMultipleUnitTests(tests: java.util.List[TestClient.ClientUnitTest.Value]): Unit = {
+    testSessionInitialization()
+    tests.asScala.map(unitTests.get).foreach(test => test.foreach(_()))
+    testShutdown()
+  }
+
+  private def testIfSuccessful[T](value: CompletableFuture[T]): T = {
+    val result = Await.ready(value.toScala, timeoutDuration).value.get
```
[the remark this answers is not in the document]
Comment (andrefmrocha) [reply testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala]: I've made it asynchronous but I was wondering if it made it sense for that to be optional. What do you think?

File: testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+  }
+
+  private def compileSuccessfully(): Unit = targetsCompileSuccessfully()
+
+  @tailrec
+  private def obtainExpectedDiagnostic(
+      expectedDiagnostic: ExpectedDiagnostic,
+      list: List[PublishDiagnosticsParams] = List.empty
+  ): List[PublishDiagnosticsParams] = {
+    val polledDiagnostic = client.pollPublishDiagnostics(alreadySentDiagnosticsTimeout)
```
[the remark this answers is not in the document]
Comment (andrefmrocha) [reply testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala]: Let me know what you think of the new solution!

[the remark this answers is not in the document]
Comment (jastice) [reply testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala]: looks good!

File: testkit/src/main/scala/ch/epfl/scala/bsp/testkit/mock/HappyMockServer.scala [hunk 6]
```diff
@@ -71,223 +106,388 @@ class HappyMockServer(base: File) extends AbstractMockServer {
...
+    }
   }
 
-  override def buildTargetRun(params: RunParams): CompletableFuture[RunResult] = {
-    // TODO some task notifications
-    val result = new RunResult(StatusCode.OK)
-    result.setOriginId(params.getOriginId)
-    completeFuture(result)
+  private def getValue[T](future: CompletableFuture[Either[T, ResponseError]]): CompletableFuture[T] = {
+    Try(future.get()) match {
```
Comment (jastice) [inline testkit/src/main/scala/ch/epfl/scala/bsp/testkit/mock/HappyMockServer.scala:467]: function returns a CompletableFuture, so waiting here is not necessary, you can chain instead.

File: testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala [hunk 7]
```diff
@@ -0,0 +1,589 @@
...
+        val targetsId = targets.map(_.getId).asJava
+        val buildTargetSources =
+          session.connection.server.buildTargetSources(new SourcesParams(targetsId)).toScala
+        val dependencySources =
+          session.connection.server
+            .buildTargetDependencySources(new DependencySourcesParams(targetsId))
+            .toScala
+        val resources =
+          session.connection.server.buildTargetResources(new ResourcesParams(targetsId)).toScala
+        Await.result(
```
Comment (jastice) [inline testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala:251]: don't await within a future callback

File: testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala [hunk 8]
```diff
@@ -0,0 +1,589 @@
...
+      })
+  }
+
+  private def testShutdown(session: MockSession, cleanup: () => Unit): Future[Unit] = {
+    session.connection.server
+      .buildShutdown()
+      .toScala
+      .map(_ => {
+        val failedRequest =
+          Await
```
Comment (jastice) [inline testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala:226]: don't await within a future callback

File: testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala [hunk 9]
```diff
@@ -0,0 +1,589 @@
...
+            session.connection.server
+              .buildTargetTest(new TestParams(testableTargets.map(_.getId).asJava))
+          )
+        if (untestableTargets.nonEmpty)
+          futures += testIfFailure(
+            session.connection.server
+              .buildTargetTest(new TestParams(untestableTargets.map(_.getId).asJava))
+          )
+
+        Await.result(Future.sequence(futures), timeoutDuration.*(futures.size))
```
Comment (jastice) [inline testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala:302]: don't await within a future callback

File: testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala [hunk 10]
```diff
@@ -0,0 +1,589 @@
...
+      .workspaceBuildTargets()
+      .toScala
+      .map(targets => {
+        val runResultsFuture = runTargets(
+          targets.getTargets.asScala,
+          session
+        )
+        runResultsFuture.map(
+          runResults =>
+            runResults.map(runResult => {
```
Comment (jastice) [inline testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala:433]: nested maps better written as for comprehension

**Comment by andrefmrocha** [comment]
@jastice The only await now is at the boundary of the test, let me know what you think 😄 !

File: testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala [hunk 11]
```diff
@@ -366,14 +366,15 @@ class TestClient(
     session.connection.server
       .workspaceBuildTargets()
       .toScala
-      .map(targets => {
-        compileTarget(targets.getTargets.asScala, session).map(compileResult => {
+      .flatMap(targets => {
```
Comment (jastice) [inline testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala:369]: style: can do with less braces

**Comment by jvican** [comment]
Thank you for the PR, LGTM! 👍

**Pull request merged**